File: exercise2_soln.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.linear_model import LinearRegression


###### from previous exercise ######
df = pd.read_csv('./data/MX_Dengue_trends.csv')
df['Date'] = pd.to_datetime(df['Date'])


df_train = df.iloc[:36, :]

mod = LinearRegression()
mod.fit(df_train[['dengue']], df_train['Dengue CDC'])
m = mod.coef_[0]
b = mod.intercept_


df_valid = df.iloc[36:, :]
pred_static = df_valid['dengue'] * m + b


########### Part 2 ############

# (a)
pred_dyn = np.zeros(df_valid.shape[0])

for month in range(36, df.shape[0]):
    
    # define training set
    start = month - 36
    end = month
    df_train = df.iloc[start:end]
    df_pred = df.iloc[month:month+1]
    
    # fit model
    lr = LinearRegression()
    lr.fit(df_train[['dengue']], df_train['Dengue CDC'])
    
    # make prediction
    pred = lr.predict(df_pred[['dengue']])
    pred_dyn[month - 36] = pred
    

# (b)
plt.plot(df_valid['Date'], df_valid['Dengue CDC'], c='k', label='Cases')
plt.plot(df_valid['Date'], pred_static, c='r', label='Static')
plt.plot(df_valid['Date'], pred_dyn, c='g', label='Dynamic')
plt.legend()


# (c)
def rmse(truth, pred):
    return np.sqrt(np.mean((truth - pred) ** 2))

print('Static ', rmse(df_valid['Dengue CDC'], pred_static))
print('Dynamic ', rmse(df_valid['Dengue CDC'], pred_dyn))


# (d)
pred_dyn_all = np.zeros(df_valid.shape[0])
features = ['dengue', 'sintomas de dengue', 'mosquito', 'dengue sintomas']

for month in range(36, df.shape[0]):
    
    # define training set
    start = month - 36
    end = month
    df_train = df.iloc[start:end]
    df_pred = df.iloc[month:month+1]
    
    # fit model
    lr = LinearRegression()
    lr.fit(df_train[features], df_train['Dengue CDC'])
    
    # make prediction
    pred = lr.predict(df_pred[features])
    pred_dyn_all[month - 36] = pred

plt.plot(df_valid['Date'], df_valid['Dengue CDC'], c='k', label='Cases')
plt.plot(df_valid['Date'], pred_static, c='r', label='Static')
plt.plot(df_valid['Date'], pred_dyn, c='g', label='Dynamic')
plt.plot(df_valid['Date'], pred_dyn_all, c='b', label='Dynamic all vars')
plt.legend()

print('Static ', rmse(df_valid['Dengue CDC'], pred_static))
print('Dynamic ', rmse(df_valid['Dengue CDC'], pred_dyn))
print('Dynamic all vars', rmse(df_valid['Dengue CDC'], pred_dyn_all))


# (e)

# add AR terms
df['ar1'] = df['Dengue CDC'].shift(1)
df['ar2'] = df['Dengue CDC'].shift(2)

pred_dyn_ar = np.zeros(df_valid.shape[0])
features = ['dengue', 'sintomas de dengue',
            'mosquito', 'dengue sintomas', 'ar1', 'ar2']

for month in range(36, df.shape[0]):
    
    # define training set
    start = max(month - 36, 2)      # notice difference here
    end = month
    df_train = df.iloc[start:end]
    df_pred = df.iloc[month:month+1]
    
    # fit model
    lr = LinearRegression()
    lr.fit(df_train[features], df_train['Dengue CDC'])
    
    # make prediction
    pred = lr.predict(df_pred[features])
    pred_dyn_ar[month - 36] = pred

plt.plot(df_valid['Date'], df_valid['Dengue CDC'], c='k', label='Cases')
plt.plot(df_valid['Date'], pred_static, c='r', label='Static')
plt.plot(df_valid['Date'], pred_dyn, c='g', label='Dynamic')
plt.plot(df_valid['Date'], pred_dyn_all, c='b', label='Dynamic all vars')
plt.plot(df_valid['Date'], pred_dyn_ar, c='m', label='ARGO')
plt.legend()

print('Static ', rmse(df_valid['Dengue CDC'], pred_static))
print('Dynamic ', rmse(df_valid['Dengue CDC'], pred_dyn))
print('Dynamic all vars', rmse(df_valid['Dengue CDC'], pred_dyn_all))
print('ARGO', rmse(df_valid['Dengue CDC'], pred_dyn_ar))

File: exercise2_soln.R
library(lubridate)

setwd('/Users/fredlu/repos/SISMID23')


####### From previous exercise #######
################ (a) ################

df <- read.csv('./data/MX_Dengue_trends.csv')
df$Date <- as.Date(df$Date)
plot(df$Date, df$Dengue.CDC, type='l')

cor(df$Dengue.CDC, df[, 3:6])    # correlations

df_train <- df[1:36,]

mod <- lm(Dengue.CDC ~ dengue, data=df_train)
b <- mod$coefficients[1]    # intercept
m <- mod$coefficients[2]    # slope

df_valid <- df[37:nrow(df), ]
pred_static <- df_valid$dengue * m + b



################ Part 2 (a) ################


###### dynamic training ######

# empty array to store predictions
pred_dyn <- numeric(nrow(df_valid))

# sliding 36-month training window
for (month in 37:nrow(df)) {
  
  # define training set
  train_start <- month - 36
  train_end <- month - 1
  df_train <- df[train_start:train_end,]
  
  # fit model
  mod <- lm(Dengue.CDC ~ dengue, data=df_train)
  
  # add prediction using model
  pred_dyn[month - 36] <- predict(mod, df[month,])
}


################ (b) ################

plot(df_valid$Date, df_valid$Dengue.CDC, type='l', col='black')
lines(df_valid$Date, pred_static, type='l', col='red')
lines(df_valid$Date, pred_dyn, type='l', col='green')
legend(df_valid$Date[1], 25000,
       legend=c('Dengue CDC', 'Static', 'Dynamic'),
       col=c('black', 'red', 'green'), lty=1)


################ (c) ################

# how to evaluate our predictions?
rmse <- function(truth, pred) {
  sqrt(mean((truth - pred)^2))
}

rmse(df_valid$Dengue.CDC, pred_static)
rmse(df_valid$Dengue.CDC, pred_dyn)


################ (d) ################

###### dynamic training + more search terms ######

# empty array to store predictions
pred_dyn_all <- numeric(nrow(df_valid))

# sliding 36-month training window
for (month in 37:nrow(df)) {
  
  # define training set
  train_start <- month - 36
  train_end <- month - 1
  df_train <- df[train_start:train_end,]
  
  # fit model
  mod <- lm(Dengue.CDC ~ dengue + sintomas.de.dengue + mosquito + dengue.sintomas,
            data=df_train)
  
  # add prediction using model
  pred_dyn_all[month - 36] <- predict(mod, df[month,])
}

plot(df_valid$Date, df_valid$Dengue.CDC, type='l', col='black')
lines(df_valid$Date, pred_static, type='l', col='red')
lines(df_valid$Date, pred_dyn, type='l', col='green')
lines(df_valid$Date, pred_dyn_all, type='l', col='blue')
legend(df_valid$Date[1], 25000,
       legend=c('Dengue CDC', 'Static', 'Dynamic', 'Dynamic all vars'),
       col=c('black', 'red', 'green', 'blue'), lty=1)

rmse(df_valid$Dengue.CDC, pred_static)
rmse(df_valid$Dengue.CDC, pred_dyn)
rmse(df_valid$Dengue.CDC, pred_dyn_all)


################ (e) ################

###### add AR terms ######
library(dplyr)

add_ar <- function(df) {
  tmp = df %>% mutate(AR1=lag(Dengue.CDC, 1),
                      AR2=lag(Dengue.CDC, 2))
  
}

df_ar <- add_ar(df)


# empty array to store predictions
pred_dyn_ar <- numeric(nrow(df_valid))

# sliding 36-month training window
for (month in 37:nrow(df_ar)) {
  
  # define training set
  train_start <- max(month - 36, 3)    # first 2 rows of df have NAs because of AR terms
  train_end <- month - 1
  df_train <- df_ar[train_start:train_end,]
  
  
  # fit model
  mod <- lm(Dengue.CDC ~ dengue + sintomas.de.dengue + mosquito + dengue.sintomas + AR1 + AR2,
            data=df_train)
  
  # add prediction using model
  pred_dyn_ar[month - 36] <- predict(mod, df_ar[month,])
}


plot(df_valid$Date, df_valid$Dengue.CDC, type='l', col='black')
lines(df_valid$Date, pred_static, type='l', col='red')
lines(df_valid$Date, pred_dyn, type='l', col='green')
lines(df_valid$Date, pred_dyn_all, type='l', col='blue')
lines(df_valid$Date, pred_dyn_ar, type='l', col='purple')
legend(df_valid$Date[1], 25000,
       legend=c('Dengue CDC', 'Static', 'Dynamic', 'Dynamic all vars', 'ARGO'),
       col=c('black', 'red', 'green', 'blue', 'purple'), lty=1)

rmse(df_valid$Dengue.CDC, pred_static)
rmse(df_valid$Dengue.CDC, pred_dyn)
rmse(df_valid$Dengue.CDC, pred_dyn_all)
rmse(df_valid$Dengue.CDC, pred_dyn_ar)
